### PR TITLE
[Developer] Adjust alfalfa_utility

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

- Adjustments for #5236

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/utilities/alfalfa/AlfalfaActuator.hpp
Comment (jmarrec) [inline src/utilities/alfalfa/AlfalfaActuator.hpp:6]: Logger was included everywhere but not used. Because you throw std::runtime_error.

Maybe you should use a proper Logger and use LOG_AND_THROW, one of the two.

File: src/utilities/filetypes/AlfalfaJSON.cpp
Comment (jmarrec) [inline src/utilities/filetypes/AlfalfaJSON.cpp:34]: TODO

File: src/utilities/filetypes/AlfalfaJSON.cpp
Comment (jmarrec) [inline src/utilities/filetypes/AlfalfaJSON.cpp:38]: TODO

File: src/utilities/alfalfa/AlfalfaPoint_Impl.hpp
Comment (jmarrec) [inline src/utilities/alfalfa/AlfalfaPoint_Impl.hpp:20]: TODO

File: src/utilities/filetypes/AlfalfaJSON.hpp
Comment (jmarrec) [inline src/utilities/filetypes/AlfalfaJSON.hpp:102]: static a few methods

File: src/utilities/filetypes/AlfalfaJSON.cpp
Comment (TShapinsky) [inline src/utilities/filetypes/AlfalfaJSON.cpp:93]: I'm not familiar with this method of constructing, this is just a way to construct the class inline in the stack?

**Review by TShapinsky — APPROVED** [review]
Thanks for this @jmarrec just a couple questions if you have time.

File: python/engine/test/PythonEngine_GTest.cpp
Comment (jmarrec) [reply python/engine/test/PythonEngine_GTest.cpp:162]: Nothing, it's an optional arbitrary string, at most 16 characters long. I like to use it, when this is a json thing I put json, when it's a sql query I put sql... etc.

(technically speaking you could configure your editor to pick that up and apply a different syntax highlighting based on it)

https://en.cppreference.com/w/cpp/language/string_literal

```
 R"d-char-seq ﻿(optional)(r-char-seq ﻿(optional))d-char-seq ﻿(optional)"
```

> d-char-seq 	- 	A sequence of one or more d-char ﻿s, at most 16 characters long
> d-char 	- 	A character from the [basic character set](https://en.cppreference.com/w/cpp/language/charset#Basic_character_set), except parentheses, backslash and [spaces](https://en.cppreference.com/w/cpp/string/byte/isspace)

File: src/utilities/alfalfa/AlfalfaActuator.cpp
Comment (jmarrec) [reply src/utilities/alfalfa/AlfalfaActuator.cpp:3]: That's the way all other files are as far as I can tell. It's debatable whether it's a good idea, but I'd prefer consistency.

(Bit of context: We don't move things often (if at all), and I am not afraid of spending 15 minutes to write a python script with a regex or two to fixup the paths if we need to move things)

File: src/utilities/alfalfa/AlfalfaComponent.cpp
Comment (jmarrec) [reply src/utilities/alfalfa/AlfalfaComponent.cpp:6]: yes, it's a POD type [1] that'll fit in a register. I tried adding a clang-tidy check for that in LLVM (and someone else did independently) but it never got in.

[1] This may be an unsigned int, and it is on GCC for eg (cf https://gcc.godbolt.org/z/47z5z3bP8), but it's actually implementation defined per the cpp standards.

File: src/utilities/alfalfa/AlfalfaPoint.cpp
Comment (jmarrec) [reply src/utilities/alfalfa/AlfalfaPoint.cpp:106]: boost::optional has an `operator bool()`, which is going to be used here.

FYI, C++17 added init-statement to `if`. So you could do something like

`if (auto x = func(); x.condition()) {`

https://en.cppreference.com/w/cpp/language/if

File: src/utilities/alfalfa/AlfalfaPoint.cpp
Comment (jmarrec) [reply src/utilities/alfalfa/AlfalfaPoint.cpp:13]: this is a global static constexpr variable. I could omit the `static`, and it'd still be `static` because it's declared at global scope and it's constexpr, but this is more explicit.

And so it initializes it in static storage (the `.rodata` section of the binary). The string_view is materialized in the binary.

see this short example: https://gcc.godbolt.org/z/o7veTMETP

and if you want to learn more about constexpr, I suggest C++ weekly, the start of this episode maybe: https://www.youtube.com/watch?v=IDQ0ng8RIqs

File: src/utilities/filetypes/AlfalfaJSON.cpp
Comment (jmarrec) [reply src/utilities/filetypes/AlfalfaJSON.cpp:93]: This is by definition a temporary (because it does not have a name), so it's an R-value, and I know that there will be no copies.

In this specific case, this is unnecessary, because the `exposeFromComponent` actually takes a const-ref, so no copy or move will happen, but that may change... **It didn't hinder readability much here**, so I just went with it.

This is heavily related to the pass-by-value and move idiom.

Here is something to play for you, switch the defines on line 41 and 42: https://gcc.godbolt.org/z/911xK1bT6

You'll see that if exposeFromComponent takes by-value:
* If you NAME your variable, the Ctor is called, then the Copy Ctor
* If you use a temporary (like I did here), the variable is actually constructed directly in the exposeFromComponent scope and you save a copy.

File: src/utilities/alfalfa/AlfalfaActuator.cpp
Comment (jmarrec) [reply src/utilities/alfalfa/AlfalfaActuator.cpp:35]: https://live.boost.org/doc/libs/1_79_0/libs/optional/doc/html/boost_optional/reference/header__boost_optional_optional_hpp_/detailed_semantics___optional_values.html

image

File: src/utilities/alfalfa/AlfalfaPoint_Impl.hpp
Comment (jmarrec) [reply src/utilities/alfalfa/AlfalfaPoint_Impl.hpp:20]: It's annoying to write for sure. For ModelObjects we actually have ruby scripts that will do 90% of the work for us and generate the pimpl class: https://github.com/NREL/OpenStudio/blob/develop/developer/ruby/GenerateClass.rb and https://github.com/NREL/OpenStudio/blob/develop/developer/doc/Adding_Model_Object_to_OpenStudio.docx

I had written this when I first touched the OpenStudio codebase: https://github.com/jmarrec/OpenStudio-Tutorial-for-Cpp-Noobs?tab=readme-ov-file#pimpl

a few years later, I think the pImpl pattern in the case of model objects at least is useful for us because:

* The Impl methods are not exposed to SWIG (ruby, python, C# bindings)
* Technically you can change the Impl header without changing the public header, which is good because:
    * If a class never included the Impl header (and we have to do that too often for casting especially), then there's a net compilation time benefit
    * It forces us to treat the public header as a sacred API, and non-backward compatible changes must be really  motivated. We often remove methods from the Impl when they become moot, and we'll put a `[[deprecated]]` method in the public side for 2 or 3 releases with a clear message that it'll be removed.

**Comment by jmarrec** [comment]
@TShapinsky I took some time to give (probably too) detailed explanations in the hope that it's useful context and references.

Anyways, you're responsible for the PR / branch I'm pointing this to, so it's up to you to merge this one or not (I don't know if you have merge rights though, but I can click the button if you ask me to)

File: src/utilities/alfalfa/AlfalfaPoint_Impl.hpp
Comment (TShapinsky) [reply src/utilities/alfalfa/AlfalfaPoint_Impl.hpp:20]: That makes sense. The reason I used the pImpl pattern here was to allow the point object a user is given after creation and the point object in the points vector within `AlflalfaJSON` to all point to the same underlying data. This seemed like the best option outside of dealing with just raw or lightly wrapped pointers. If there is a better paradigm you have in mind for this I would be happy to implement it.

File: src/utilities/alfalfa/AlfalfaActuator.cpp
Comment (TShapinsky) [reply src/utilities/alfalfa/AlfalfaActuator.cpp:3]: that sounds good, currently there a question on the other PR from kyle as to whether `alfalfa` should be moved out of `utilities` to be on the same level. Ideally I think I'd wait for that before switching everything back to relative, but as you noted it's not that big of a lift to fix.

**Pull request merged**
